src/envs/doors.py

```diff
@@ -0,0 +1,933 @@
+"""A 2D navigation environment with obstacles, rooms, and doors."""
+
+import itertools
+from typing import ClassVar, Dict, Iterator, List, Optional, Sequence, Set, \
+    Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+from gym.spaces import Box
+from numpy.typing import NDArray
+
+from predicators.src import utils
+from predicators.src.envs import BaseEnv
+from predicators.src.settings import CFG
+from predicators.src.structs import Action, Array, GroundAtom, Image, Object, \
+    ParameterizedOption, Predicate, State, Task, Type
+from predicators.src.utils import Rectangle, _Geom2D
+
+
+class DoorsEnv(BaseEnv):
+    """A 2D navigation environment with obstacles, rooms, and doors."""
+    room_size: ClassVar[float] = 1.0
+    hallway_width: ClassVar[float] = 0.25
+    wall_depth: ClassVar[float] = 0.01
+    robot_radius: ClassVar[float] = 0.05
+    action_magnitude: ClassVar[float] = 0.05
+    robot_initial_position_radius: ClassVar[float] = 0.05
+    obstacle_initial_position_radius: ClassVar[float] = 0.1
+    obstacle_size_lb: ClassVar[float] = 0.05
+    obstacle_size_ub: ClassVar[float] = 0.15
+    doorway_pad: ClassVar[float] = 1e-3
+    move_sq_dist_tol: ClassVar[float] = 1e-5
+    open_door_thresh: ClassVar[float] = 1e-2
+
+    def __init__(self) -> None:
+        super().__init__()
+        # Types
+        self._robot_type = Type("robot", ["x", "y"])
+        self._door_type = Type("door", [
+            "x", "y", "theta", "mass", "friction", "rot", "target_rot", "open"
+        ])
+        self._room_type = Type("room", ["x", "y"])
+        self._obstacle_type = Type("obstacle",
+                                   ["x", "y", "width", "height", "theta"])
+        # Predicates
+        self._InRoom = Predicate("InRoom", [self._robot_type, self._room_type],
+                                 self._InRoom_holds)
+        self._InDoorway = Predicate("InDoorway",
+                                    [self._robot_type, self._door_type],
+                                    self._InDoorway_holds)
+        self._InMainRoom = Predicate("InMainRoom",
+                                     [self._robot_type, self._room_type],
+                                     self._InMainRoom_holds)
```

Comment:
I don't think I understand why we need this when we already have `self._InRoom`?